### PR TITLE
dht-store (BEP 44)

### PR DESCRIPTION
**Pull request by substack**
This gigantic patch adds support for [BEP 44: storing arbitrary data on the DHT](http://bittorrent.org/beps/bep_0044.html).

There are lots of tests, but there could always be more.
It would also be really good to make sure this works with some other clients that implement the spec, but I don't want to delay this patch any longer since it's already pretty huge.

https://github.com/feross/bittorrent-dht/issues/60

### CODE REVIEW

File: client.js [hunk 1]
```diff
@@ -259,6 +264,322 @@ DHT.prototype.announce = function (infoHash, port, cb) {
...
+    return null
+  }
+  if (isMutable && opts.seq === undefined) {
+    cb([ new Error('opts.seq not provided for a mutable update') ])
+    return null
+  }
+  if (isMutable && typeof opts.seq !== 'number') {
+    cb([ new Error('opts.seq not an integer') ])
+    return null
+  }
```
Comment (mafintosh) [inline client.js]: shouldn't it just throw on all of the above errors as they are config errors?

**Comment by mafintosh** [comment]
Impressive work @substack ! Other than my above comment, LGTM from me. @feross any comments?

**Comment by substack** [comment]
Thanks for the suggestion @mafintosh. Patch updated.

**Comment by substack** [comment]
The test vectors in bep44 give different values than I'm getting currently, so this patch will require some more changes to interoperate with other clients.

**Comment by lmatteis** [comment]
I've tested this using this code:

```
var DHT = require('bittorrent-dht')

var readline = require('readline')
var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
})

if(process.argv.length == 3) {
  var dht = new DHT({ bootstrap: false })
  dht.listen(parseInt(process.argv[2], 10), function () {
    console.log('now listening on '+ process.argv[2]+ ' as bootstrap')
  })   
} else {
  var dht = new DHT({ bootstrap: [process.argv[2]] })   
  dht.listen(parseInt(process.argv[3], 10), function () {
    console.log('now listening on '+process.argv[3]+' with bootstrap '+ process.argv[2])
  })  
}

dht.on('ready', function () {
  rl.on('line', function(line){
    var s = line.split(' ')
    var cmd = s[0]
    if(cmd == 'get') {
      var val = new Buffer(s[1], 'hex')
      dht.get(val, function(err, res) {
        if(err) return console.error(err)
        console.log(res.v.toString())
      })  
    } else if(cmd == 'put') {
      var val = new Buffer(s[1], 'utf-8')
      dht.put({ v: val }, function (errors, hash) {
        console.error('errors=', errors)
        console.log('hash=', hash.toString('hex'))
      })  
    }   
  })  
})
```

I run 3 processes of this and they all communicate fine:

```
$ node main.js 3000
now listening on 3000 as bootstrap
put beep 
errors= []
hash= 38d8fe5c3796e1b93ceb59dcdb879c010507e64a

$ node main.js 127.0.0.1:3000 3001
now listening on 3001 with bootstrap 127.0.0.1:3000
get 38d8fe5c3796e1b93ceb59dcdb879c010507e64a
beep

$ node main.js 127.0.0.1:3001 3002
now listening on 3002 with bootstrap 127.0.0.1:3001
get 38d8fe5c3796e1b93ceb59dcdb879c010507e64a
beep
```

Problem is when a new node joins the network, it doesn't find `beep`:

```
$ node main 127.0.0.1:3002 3003
now listening on 3003 with bootstrap 127.0.0.1:3002
get 38d8fe5c3796e1b93ceb59dcdb879c010507e64a
[Error: hash not found]
```

Any ideas what it could be? Thanks!

**Comment by cjb** [comment]
@lmatteis Hey, did you try running with `DEBUG=*`?  If I do that, then your code in the failing case gives:
```
  bittorrent-dht [49c9e95] Invalid data: Missing delimiter ":" [0x3a] from 192.34.86.36:30000 (d1:rd2:id20:]��^f�?4�n�y�B���0c�1:k5:nodesle6:nodes6le4:salt3:seq3:sig5:token1:v4:beepe1:t2:1:y1:re) +29ms


[Error: hash not found]
```
.. but in the passing case it doesn't happen, suggesting that the data's only invalid if its "put" wasn't already listened to, which is strange.  Could you confirm you're seeing the same?

**Comment by cjb** [comment]
@substack @feross If you're able to take a look, I think this is the problem:

We have test cases for:
```
node 1 joins
node 2 joins
node 1 performs put() and broadcasts it to node 2
node 2 performs get()
node 2 answers itself using the dht-store value that it cached in `self.nodes[node].data`
```
.. but we don't have test cases for:
```
node 1 joins
node 1 performs put()
node 2 joins
node 2 performs get()
node 1 sends an answer to the get() request
```
.. and when this happens, node 2 doesn't understand node 1's answer -- it fails to bencode.decode().

So we support get() from a put() that was observed by the querying node, but not get() from an unobserved put().

I'll keep looking into what the problem is, help would be super appreciated :)

This isn't a helpful format, but here's another example of a get() response that fails bdecoding:
```
  bittorrent-dht [9272143] Invalid data: Missing delimiter ":" [0x3a] from 72.226.1.185:6882 (d1:rd2:id20:8f�x����4m����m��G�1:k32:$R��k?��s~6'lH���f��L�'�l����5:nodesle6:nodes6le4:salt3:seqi0e3:sig64:t%$YѰ,����5i��W���D\z�T��lmne�(��.����5:token20:�̓"ZsP��eW���.��݋Y1:v20:abcabcabcabcabcabcabe1:t2:1:y1:re) +22ms
```

**Comment by lmatteis** [comment]
Hi @cjb - here's my full output:

```
get 38d8fe5c3796e1b93ceb59dcdb879c010507e64a
  bittorrent-dht [ffc6fc7] lookup (get_peers) 38d8fe5c3796e1b93ceb59dcdb879c010507e64a +20s
  bittorrent-dht [ffc6fc7] sent get_peers 38d8fe5c3796e1b93ceb59dcdb879c010507e64a to 127.0.0.1:3000 +1ms
  bittorrent-dht [ffc6fc7] sent get_peers 38d8fe5c3796e1b93ceb59dcdb879c010507e64a to 127.0.0.1:3001 +0ms
  bittorrent-dht [ffc6fc7] sent get_peers 38d8fe5c3796e1b93ceb59dcdb879c010507e64a to 127.0.0.1:3000 +1ms
  bittorrent-dht [ffc6fc7] addNode f3d047100928e8986f6d522d07154e255e3534c9 127.0.0.1:3001 discovered from 127.0.0.1:3001 +2ms
  bittorrent-dht [ffc6fc7] addNode 38d8fe5c3796e1b93ceb59dcdb879c010507e64a 127.0.0.1:3000 discovered from 127.0.0.1:3001 +1ms
  bittorrent-dht [ffc6fc7] skip addNode 127.0.0.1:3002 ffc6fc713c6356903349b9ec7debea2bf04be740; that is us! +0ms
  bittorrent-dht [ffc6fc7] addNode e8b9c6963e518768293b117eb358575153fa128d 127.0.0.1:3000 discovered from 127.0.0.1:3001 +0ms
  bittorrent-dht [ffc6fc7] got lookup response from f3d047100928e8986f6d522d07154e255e3534c9 +0ms
  bittorrent-dht [ffc6fc7] addNode e8b9c6963e518768293b117eb358575153fa128d 127.0.0.1:3000 discovered from 127.0.0.1:3000 +0ms
  bittorrent-dht [ffc6fc7] addNode 38d8fe5c3796e1b93ceb59dcdb879c010507e64a 127.0.0.1:3000 discovered from 127.0.0.1:3000 +1ms
  bittorrent-dht [ffc6fc7] skip addNode 127.0.0.1:3002 ffc6fc713c6356903349b9ec7debea2bf04be740; that is us! +0ms
  bittorrent-dht [ffc6fc7] addNode f3d047100928e8986f6d522d07154e255e3534c9 127.0.0.1:3001 discovered from 127.0.0.1:3000 +0ms
  bittorrent-dht [ffc6fc7] got lookup response from e8b9c6963e518768293b117eb358575153fa128d +0ms
  bittorrent-dht [ffc6fc7] addNode e8b9c6963e518768293b117eb358575153fa128d 127.0.0.1:3000 discovered from 127.0.0.1:3000 +0ms
  bittorrent-dht [ffc6fc7] addNode 38d8fe5c3796e1b93ceb59dcdb879c010507e64a 127.0.0.1:3000 discovered from 127.0.0.1:3000 +0ms
  bittorrent-dht [ffc6fc7] skip addNode 127.0.0.1:3002 ffc6fc713c6356903349b9ec7debea2bf04be740; that is us! +0ms
  bittorrent-dht [ffc6fc7] addNode f3d047100928e8986f6d522d07154e255e3534c9 127.0.0.1:3001 discovered from 127.0.0.1:3000 +0ms
  bittorrent-dht [ffc6fc7] got lookup response from e8b9c6963e518768293b117eb358575153fa128d +0ms
  bittorrent-dht [ffc6fc7] terminating lookup (get_peers) 38d8fe5c3796e1b93ceb59dcdb879c010507e64a +1ms
  bittorrent-dht [ffc6fc7] K closest nodes are: +0ms
  bittorrent-dht [ffc6fc7]   127.0.0.1:3001 f3d047100928e8986f6d522d07154e255e3534c9 +0ms
  bittorrent-dht [ffc6fc7]   127.0.0.1:3000 e8b9c6963e518768293b117eb358575153fa128d +0ms
  bittorrent-dht [ffc6fc7] Invalid data: Missing delimiter ":" [0x3a] from 127.0.0.1:3001 (d1:rd2:id20:��G      (��omR-N%^54�1:k5:nodesle6:nodes6le4:salt3:seq3:sig5:token1:v4:beepe1:t2:1:y1:re) +1ms
  bittorrent-dht [ffc6fc7] Invalid data: Missing delimiter ":" [0x3a] from 127.0.0.1:3000 (d1:rd2:id20:��Ɩ>Q�h);~�XWQS��1:k5:nodesle6:nodes6le4:salt3:seq3:sig5:token1:v4:beepe1:t2:1:y1:re) +1ms
[Error: hash not found]
```

**Comment by cjb** [comment]
@lmatteis Cool.  So you're confirming that you *are* getting sent the "beep" key that you want, and the client just isn't understanding the result, so it thinks it didn't get anything back.  We need to figure out why this result is failing to `bencode.decode()` and fix it.

**Comment by cjb** [comment]
I've fixed this for immutable hashes now, thanks to @lmatteis for looking at it with me.  The fixes are on my branch here: https://github.com/substack/bittorrent-dht/compare/dht-store...cjb:dht-store?expand=1

There were three problems:

1) we sent the mutable key fields back to immutable key requests in _onGet(), but the values for those fields (k/salt/seq/sig/cas) were undefined, and

2) it turns out when you bencode.encode() an object with an undefined value like this, encode() doesn't complain but then bencode.decode(bencode.encode(..)) fails!  Maybe we should decode() the things we're about to send out on the wire to check for this?

3) the immutable hash check had broken logic, it did:
```
          } else if (!isMutable && bufcmp(sha1(res.v), hash)) {
            self._debug('invalid immutable hash from %s', node.addr)
```
instead of
```
          } else if (!isMutable && !bufcmp(sha1(res.v), hash)) {
            self._debug('invalid immutable hash from %s', node.addr)
```
Moving on to fixing mutable hash get now.  Help welcome!

**Comment by cjb** [comment]
My branch has both immutable and mutable hash lookup working now:

https://github.com/substack/bittorrent-dht/pull/1

(There's still an issue that @substack mentioned on IRC with the key format and interoperability with non-bittorrent-dht clients.)

**Comment by substack** [comment]
@mvayngrib I get the same value every time I run the tests, so I think it's something else. The folks in #libtorrent agree that it's a difference in key formatting.

**Comment by mvayngrib** [comment]
re: same sig every time, you're right, i just tried it myself. My bad :)

re: formatting difference, definitely. The sigs in the test vectors are 64 bytes, so just r and s (padded if needed). Elliptic's sigs are 70 bytes:

[ 0x30, length byte, 0x02, r length byte, r, 0x02, s length byte, s]

**Comment by maxogden** [comment]
bump!

**Comment by substack** [comment]
This patch is now using [ed25519-supercop](https://github.com/substack/ed25519-supercop) and the test vectors match the spec. Ready to merge.

**Comment by substack** [comment]
On further inspection, there is an issue with the signature data that didn't surface in the tests. Patch coming soon.

**Comment by substack** [comment]
Patch all good to go! I'm not sure what's up with the node-gyp issue on ed25519-supercop in the travis builds.

**Comment by substack** [comment]
travis issue sorted out using the .travis.yml from leveldown for reference

**Comment by mafintosh** [comment]
:+1: from me. @feross travis is green! LGTM

**Comment by mafintosh** [comment]
Released as 4.0.0. I'm major bumping this just to be safe, as this is a fairly big patch.

**Pull request merged**